ArduCopter/Attitude.cpp

```diff
@@ -1,21 +1,285 @@
 #include "Copter.h"
+#include <AP_HAL/CondMutex.h>
 
 /*************************************************************
  *  Attitude Rate controllers and timing
  ****************************************************************/
 
-// update rate controllers and output to roll, pitch and yaw actuators
-//  called at 400hz by default
-void Copter::run_rate_controller()
+#if AP_INERTIALSENSOR_RATE_LOOP_WINDOW_ENABLED
+
+//#define RATE_LOOP_TIMING_DEBUG
+/*
+  thread for rate control
+*/
+void Copter::rate_controller_thread()
+{
+    uint8_t rate_decimation = 1;
+    uint32_t rate_loop_count = 0;
+    uint32_t prev_loop_count = 0;
+
+    HAL_CondMutex cmutex;
+    ins.set_rate_loop_mutex(&cmutex);
+
+    uint32_t last_run_us = AP_HAL::micros();
+    float max_dt = 0.0;
+    float min_dt = 1.0;
+    uint32_t now_ms = AP_HAL::millis();
+    uint32_t last_rate_check_ms = 0;
+    uint32_t last_rate_increase_ms = 0;
+#if HAL_LOGGING_ENABLED
+    uint32_t last_rtdt_log_ms = now_ms;
+#endif
+    uint32_t last_notch_sample_ms = now_ms;
+    bool was_using_rate_thread = false;
+    uint32_t running_slow = 0;
+#ifdef RATE_LOOP_TIMING_DEBUG
+    uint32_t gyro_sample_time_us = 0;
+    uint32_t rate_controller_time_us = 0;
+    uint32_t motor_output_us = 0;
+    uint32_t timing_count = 0;
+    uint32_t last_timing_msg_us = 0;
+#endif
+
+    uint8_t filter_rate_decimate = 2;
+    if (copter.g2.att_filter_rate_hz != 0) {
+        filter_rate_decimate = uint8_t((ins.get_raw_gyro_rate_hz() + copter.g2.att_filter_rate_hz - 1) / copter.g2.att_filter_rate_hz);
+    }
+    uint8_t log_rate_decimate = 0;
+    if (copter.g2.att_log_rate_hz != 0) {
+        log_rate_decimate = uint8_t((ins.get_raw_gyro_rate_hz() + copter.g2.att_log_rate_hz - 1) / copter.g2.att_log_rate_hz);
+    }
+    uint8_t filter_loop_count = 0;
+    uint8_t log_loop_count = 0;
+
+    while (true) {
+
+#ifdef RATE_LOOP_TIMING_DEBUG
+        uint32_t rate_now_us = AP_HAL::micros();
+#endif
+
+        // allow changing option at runtime
+        if ((!flight_option_is_set(FlightOptions::USE_RATE_LOOP_THREAD) &&
+             !flight_option_is_set(FlightOptions::USE_FIXED_RATE_LOOP_THREAD)) ||
+            ap.motor_test) {
+            using_rate_thread = false;
+            if (was_using_rate_thread) {
+                // if we were using the rate thread, we need to
+                // setup the notch filter sample rate
+                const float loop_rate_hz = AP::scheduler().get_filtered_loop_rate_hz();
+                attitude_control->set_notch_sample_rate(loop_rate_hz);
+                motors->set_dt(1.0/loop_rate_hz);
```

Comment:
motors dt is set lots of places, but I don't think the motors update rate is actually changed?

Reply:
Fixed